staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -346,9 +346,10 @@ func (le *LeaderElector) tryAcquireOrRenew(ctx context.Context) bool {
 
 		le.observedRawRecord = oldLeaderElectionRawRecord
 	}
-	if len(oldLeaderElectionRecord.HolderIdentity) > 0 &&
-		le.observedTime.Add(le.config.LeaseDuration).After(now.Time) &&
-		!le.IsLeader() {
```

Comment:
It looks like this is what will make the `le.release()` take effect
```go
	if len(oldLeaderElectionRecord.HolderIdentity) > 0 &&
		le.observedTime.Add(time.Second * time.Duration(oldLeaderElectionRecord.LeaseDurationSeconds).After(now.Time) &&
		!le.IsLeader() {
```

It seems to be better to determine whether `oldLeaderElectionRecord` is expired or not based on  `oldLeaderElectionRecord.LeaseDurationSeconds`, and it will not affect the skewed clock issue mentioned before.

@liggitt  what do you think?

Reply:
I think this makes sense... the previous logic would let a leader configured with a short duration think it had acquired the lease from a leader with a long duration, ignoring the LeaseDuration recorded in the object

Reply:
This needs test coverage to cover the logic changed here

Reply:
I just realized that we are not using the `clock` injected into the struct to calculate the time `Now`

```
diff --git a/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go b/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
index eb72b2517bc..213d953ef5c 100644
--- a/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
+++ b/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
@@ -292,7 +292,7 @@ func (le *LeaderElector) release() bool {
        if !le.IsLeader() {
                return true
        }
-       now := metav1.Now()
+       now := metav1.NewTime(le.clock.Now())
        leaderElectionRecord := rl.LeaderElectionRecord{
                LeaderTransitions:    le.observedRecord.LeaderTransitions,
                LeaseDurationSeconds: 1,
@@ -312,7 +312,7 @@ func (le *LeaderElector) release() bool {
 // else it tries to renew the lease if it has already been acquired. Returns true
 // on success else returns false.
 func (le *LeaderElector) tryAcquireOrRenew(ctx context.Context) bool {
-       now := metav1.Now()
+       now := metav1.NewTime(le.clock.Now())
        leaderElectionRecord := rl.LeaderElectionRecord{
                HolderIdentity:       le.config.Lock.Identity(),
                LeaseDurationSeconds: int(le.config.LeaseDuration / time.Second),
```

Reply:
@liggitt despite it makes sense, the definition of `LeaseDuration`is precisely what is doing in this case

https://github.com/kubernetes/kubernetes/blob/bb7c9739a37dabc3d70831217dc3c772348b87df/staging/src/k8s.io/component-base/config/types.go#L46-L52

I don't think we should modify this

Reply:
So, when we release we set the LeaseDuration to 1 

https://github.com/kubernetes/kubernetes/blob/ff763a144be2873552e822f7a76d1846bf918519/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L291-L301

I think that we can just detect that condition, if LeaseDuration == 1 then try to acquire it because we know the other has released it (is that a reasonable assumption)?

If LeaseDuraiton != 1 respect our configured LeaseDuration per the field definition

Reply:
> @liggitt despite it makes sense, the definition of LeaseDurationis precisely what is doing in this case

I don't see how can be correct to use the locally configured LeaseDuration and ignore the lease duration set on the Lease object. The locally configured LeaseDuration should control what the local component *sets* as the duration when it is creating/updating the lease object, but if another instance has a longer configured lease duration, surely every other component should wait for the lease duration recorded in the Lease object, not their local shorter duration?

Reply:
> I don't see how can be correct to use the locally configured LeaseDuration and ignore the lease duration set on the Lease object.

Put another way, I think using the locally configured LeaseDuration time is only correct if all instances use the same configuration.

Reply:
it makes sense, I'm trying to think if we are omitting some weird case that doesn't work, but I can't find one ... then we have to update the field doc too

Reply:
the only use case I can see to have it this way is if someone want to take the leader over a slow leader, but I feel that is a bit "aggresive"

Reply:
> if someone want to take the leader over a slow leader

taking a lease from an existing leader before the duration recorded in the Lease object expires seems indistinguishable from a leader election bug to me

Reply:
I agree with @aojea about using `le.clock` for the time calculations here, and test coverage is still required.

Reply:
@liggitt using `le.clock` and adding tests